pkg/kubelet/userns/userns_manager.go

```diff
@@ -379,20 +380,31 @@ func (m *UsernsManager) createUserNs(pod *v1.Pod) (userNs userNamespace, err err
 }
 
 // GetOrCreateUserNamespaceMappings returns the configuration for the sandbox user namespace
-func (m *UsernsManager) GetOrCreateUserNamespaceMappings(pod *v1.Pod) (*runtimeapi.UserNamespace, error) {
-	if !utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport) {
+func (m *UsernsManager) GetOrCreateUserNamespaceMappings(pod *v1.Pod, runtimeHandler string) (*runtimeapi.UserNamespace, error) {
+	featureEnabled := utilfeature.DefaultFeatureGate.Enabled(features.UserNamespacesSupport)
+
+	if pod == nil || pod.Spec.HostUsers == nil || *pod.Spec.HostUsers {
+		// if the feature is enabled, specify to use the node mode...
+		if featureEnabled {
+			return &runtimeapi.UserNamespace{
+				Mode: runtimeapi.NamespaceMode_NODE,
+			}, nil
+		}
+		// ...otherwise don't even specify it
 		return nil, nil
 	}
 
+	// if the pod requested a user namespace and the runtime doesn't support user namespaces then return an error,
+	// even when the feature gate is disabled.
+	if handlerSupportsUserns, err := m.kl.HandlerSupportsUserNamespaces(runtimeHandler); err != nil {
+		return nil, err
+	} else if !handlerSupportsUserns {
+		return nil, fmt.Errorf("the handler %q does not support user namespaces", runtimeHandler)
```

Comment:
I wonder what we should do with the pods we read from disk at initialization. If they have userns and the feature gate is disabled, what is the right action to take?

Reply:
I guess we should fail those pods too? Or what would be the right thing?

Reply:
It depends if the pod are already running. If they are, we can't kill them. If they aren't (is there any case where they aren't and we just start them?), then reject them.

Reply:
Thinking more about this, if we can read from disk a pod with userns, it was created when it was accepted (or root on the host created the files ideantically as the kubelet does, we don't care about that in our threat model), so it is okay to start them again, if that case can even exist.

If pods with userns are created, the only reliable and documented way to disallow them, is to disable the feature gate and re-create them.

So, it is completely fine to ignore this when reading the pods from disk, IMHO.

Reply:
Oh, I can't mark this as resolved myself. But it is resolved from my side :)